### PR TITLE
Sync client

### PR DESCRIPTION
**Pull request by vaind**
Adds sync client support following the same API as in line with other language bindings. Currently, tests are rather limited as they require a special sync-enabled runtime library.

### TODOs:
* [x] Requires next c-api (v0.11) for `obx_sync_available()`
* [x] sync annotation
* [x] test with an actual sync server
* [x] integrate with sync-enabled Android/iOS objectbox packages
* [x] docs updates
* [ ] [postponed] listeners

### CODE REVIEW

File: lib/src/store.dart [hunk 1]
```diff
@@ -17,6 +17,9 @@ class Store {
   Pointer<Void> _cStore;
   final ModelDefinition defs;
 
+  // a list of observers of the Store.close() event
+  final Map<dynamic, void Function()> _onBeforeClose;
+
```
Comment (vaind) [inline lib/src/store.dart:22]: @greenrobot-team & @Buggaboo please have a look at this functionality (including the remaining changes in `store.dart`). I'd like to cherry-pick that into main as it's also useful for other parts of the library, e.g. observable query `unsubscribe()` could be done automatically during store `close()`.

Comment (Buggaboo) [reply lib/src/store.dart:22]: cool!

Comment (greenrobot-team) [reply lib/src/store.dart:22]: I don't follow what
```
listeners.forEach((observer) => observer());
```
is supposed to do. Am I missing something?

Comment (vaind) [reply lib/src/store.dart:22]: executes each "onBeforeClose" event listeners, or in other words, issues the event to all listeners

Comment (greenrobot-team) [reply lib/src/store.dart:22]: 🤦 Ah, got it. I assumed this was supposed to call a close method that all observers should implement. Should have looked more closely at the type of `_onBeforeClose`... Observers will add themselves as observers to the Store close event.

LGTM!

Comment (vaind) [reply lib/src/store.dart:22]: modified that calling code slightly and added some comments to improve readability

**Comment by vaind** [comment]
@greenrobot-team  please have an overall look. I don't expect any major changes, with listeners postponed to a next issue.

File: lib/src/bindings/signatures.dart [hunk 2]
```diff
@@ -6,6 +6,19 @@ import 'structs.dart';
 
 // ignore_for_file: non_constant_identifier_names
 
+// typedefs for common signatures for different "classes", like store, box, ...
```
Comment (greenrobot-team) [inline lib/src/bindings/signatures.dart:9]: Yes, great idea! 👍

File: lib/src/annotations.dart [hunk 3]
```diff
@@ -13,14 +14,22 @@ class Entity {
...
+
   const Id({this.uid});
 }
 
 class Transient {
   const Transient();
 }
+
+/// See Sync() in sync.dart.
+/// Defining a class with the same name here would couse a duplicate export.
```
Comment (greenrobot-team) [inline lib/src/annotations.dart:32]: `couse` 😁

**Review by greenrobot-team — CHANGES_REQUESTED** [review]
Just some nitpicking comments. LGTM otherwise!

File: lib/src/sync.dart [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+      case OBXSyncState.LOGGED_IN:
+        return SyncState.loggedIn;
+      case OBXSyncState.DISCONNECTED:
+        return SyncState.disconnected;
+      case OBXSyncState.STOPPED:
+        return SyncState.stopped;
+      case OBXSyncState.DEAD:
+        return SyncState.dead;
+      default:
+        throw Exception('Unknown sync state: ' + state.toString());
```
[the remark this answers is not in the document]
Comment (vaind) [reply lib/src/sync.dart:124]: OK

File: lib/src/sync.dart [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+
+  SyncCredentials.googleAuthUint8List(this._data)
+      : _type = OBXSyncCredentialsType.GOOGLE_AUTH;
+
+  SyncCredentials.googleAuthString(String data)
+      : _type = OBXSyncCredentialsType.GOOGLE_AUTH,
+        _data = Uint8List.fromList(data.codeUnits);
+}
+
+// TODO check enum name/align with other bindings - maybe SyncClientState?
```
[the remark this answers is not in the document]
Comment (vaind) [reply lib/src/sync.dart:38]: OK

File: lib/src/sync.dart [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+  const Sync();
+
+  static bool isAvailable() {
+    return bindings.obx_sync_available() != 0;
+  }
+
+  /// Creates a sync client associated with the given store and configures it with the given options.
+  /// This does not initiate any connection attempts yet: call SyncClient::start() to do so.
+  /// Before start(), you can still configure some aspects of the sync client, e.g. its "request update" mode.
+  /// @note While you may not interact with SyncClient directly after start(), you need to hold on to the object.
```
[the remark this answers is not in the document]
Comment (vaind) [reply lib/src/sync.dart:222]: yep

File: test/sync_test.dart [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    test('SyncClient - data test (requires manual server setup)', () {
+      final env2 = TestEnv('sync2');
+
+      loggedInClient(env.store);
+      loggedInClient(env2.store);
+
+      int id = env.box.put(TestEntity(tLong: Random().nextInt(1 << 32)));
+      expect(waitUntil(() {
+        try {
+          env2.box.get(id);
```
[the remark this answers is not in the document]
Comment (vaind) [reply test/sync_test.dart:146]: yeah, 1 ms is plenty of time to sleep I hope

File: test/sync_test.dart [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+      expect(read1.id, equals(read2.id));
+      expect(read1.tLong, equals(read2.tLong));
+    },
+        // Note: only available when you start a sync server manually.
+        // Comment out the `skip: ` argument in tthe test-case definition.
+        // run sync-server --unsecured-no-authentication --model=/path/objectbox-dart/test/objectbox-model.json
+        skip: 'Data sync test is disabled, Enable after running sync-server.' //
+        );
+  } else {
+    // TESTS to run when SYNC isn't available
```
[the remark this answers is not in the document]
Comment (vaind) [reply test/sync_test.dart:164]: Don't know what you mean. This else belongs to the following `if`:

```dart
  if (Sync.isAvailable()) {
    // TESTS to run when SYNC is available
    ...
  } else {
    // TESTS to run when SYNC isn't available
   ...
  }
```

File: lib/src/annotations.dart (hunk 3 above)
Comment (vaind) [reply lib/src/annotations.dart:32]: thx

File: test/sync_test.dart [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+      expect(c2, isNotNull);
+      expect(c2.isClosed(), isFalse);
+      c2.close();
+      expect(store.syncClient(), isNull);
+
+      // closing a store closes a client
+      final env2 = TestEnv('sync2');
+      final c3 = createClient(env2.store);
+      env2.close();
+      expect(c3.isClosed(), isTrue);
```
[the remark this answers is not in the document]
Comment (vaind) [reply test/sync_test.dart]: split up

File: generator/lib/src/code_builder.dart [hunk 10]
```diff
@@ -153,10 +153,11 @@ class CodeBuilder extends Builder {
       // in case the entity is created (i.e. when its given UID or name that does not yet exist), we are done, as nothing needs to be merged
       final createdEntity = modelInfo.addEntity(entity);
       return createdEntity.id;
+    } else {
```
[the remark this answers is not in the document]
Comment (vaind) [reply generator/lib/src/code_builder.dart:156]: Restructured, please have another look

File: lib/src/sync.dart [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+  SyncCredentials.none()
+      : _type = OBXSyncCredentialsType.NONE,
+        _data = Uint8List(0);
+
+  SyncCredentials.sharedSecretUint8List(this._data)
+      : _type = OBXSyncCredentialsType.SHARED_SECRET;
+
+  SyncCredentials.sharedSecretString(String data)
+      : _type = OBXSyncCredentialsType.SHARED_SECRET,
+        _data = Uint8List.fromList(data.codeUnits);
```
[the remark this answers is not in the document]
Comment (vaind) [reply lib/src/sync.dart:28]: Not exactly sure how that should work... hopefully this is the way to go: https://stackoverflow.com/a/51101618

[the remark this answers is not in the document]
Comment (vaind) [reply lib/src/sync.dart]: Utf8.toUtf8() is an FFI interface. I've changed the code to use `utf8` from the `dart:convert` package

 :crossed_fingers: that it works

[the remark this answers is not in the document]
Comment (vaind) [reply lib/src/sync.dart]: I'll also add this to the "manual" test case

[the remark this answers is not in the document]
Comment (vaind) [reply lib/src/sync.dart]: looks like `utf8.encode()` from `dart:convert` is the way to go - produces the same byte-sequence as a server would when launched with such shared secret.

File: lib/src/store.dart (hunk 1 above)
Comment (vaind) [reply lib/src/store.dart]: I've changed this a little before merging and also took the liberty to update `observable.dart`. If you could, @Buggaboo, please have a look at 19b8008bf9269a9fa049de6952bbaeaeb1ad6aa5 if you see any issues

Comment (Buggaboo) [reply lib/src/store.dart]: Looks good!

File: test/sync_test.dart (hunk 8 above)
[the remark this answers is not in the document]
Comment (greenrobot-team) [reply test/sync_test.dart:198]: 🤦  Duh, you're right.

File: generator/lib/src/code_builder.dart (hunk 10 above)
[the remark this answers is not in the document]
Comment (greenrobot-team) [reply generator/lib/src/code_builder.dart:156]: Even better!

**Comment by greenrobot-team** [comment]
Latest changes LGTM.

**Pull request merged**